src/app/api/activity-logs/services/activity-log.service.ts

```diff
@@ -28,49 +31,88 @@ export class ActivityLogService extends BaseService {
     `
     const parsedActivityLogs = DBActivityLogArraySchema.parse(activityLogs)
 
-    // @todo fetch all users - internal as well as client and pass them to each transformer for name, profile picture etc.
+    const copilotService = new CopilotAPI(this.user.token)
+
+    const promises_getInternalUser = parsedActivityLogs.map(async (activityLog) => {
+      return copilotService.getInternalUser(activityLog.userId)
+    })
+
+    const internalUsers = await Promise.all(promises_getInternalUser)
 
     const commentIds = parsedActivityLogs
       .filter((activityLog) => activityLog.type === ActivityType.COMMENT_ADDED)
       .map((activityLog) => activityLog.details.id)
       .filter((commentId: unknown): commentId is string => commentId !== null)
 
-    // @todo move the db call to comment service
-    const comments = await this.db.comment.findMany({
-      where: {
-        id: {
-          in: commentIds,
-        },
-      },
-    })
+    const commentService = new CommentService(this.user)
+    const comments = await commentService.getCommentsByIds(commentIds)
 
-    return parsedActivityLogs.map((activityLog) => {
-      return {
-        ...activityLog,
-        details: this.formatActivityLogDetails(activityLog.type, activityLog.details, comments),
-        createdAt: activityLog.createdAt.toISOString(),
-        initiator: {
-          // @todo filter users based on the userId from the activity log and return id, name and profile picture
-        },
-      }
-    })
+    return await Promise.all(
+      parsedActivityLogs.map(async (activityLog) => {
+        return {
+          ...activityLog,
+          details: await this.formatActivityLogDetails(activityLog.type, activityLog.details, comments),
+          createdAt: activityLog.createdAt.toISOString(),
+          initiator: {
+            ...internalUsers.find((iu) => iu.id === activityLog.userId),
+          },
+        }
+      }),
+    )
   }
 
-  formatActivityLogDetails<ActivityLog extends keyof typeof SchemaByActivityType>(
+  async formatActivityLogDetails<ActivityLog extends keyof typeof SchemaByActivityType>(
     activityType: ActivityLog,
     payload: DBActivityLogDetails,
     comments: Comment[],
   ) {
+    const copilotService = new CopilotAPI(this.user.token)
     switch (activityType) {
       case ActivityType.COMMENT_ADDED:
         const comment = comments.find((comment) => comment.id === payload.id)
         if (!comment) {
           throw new Error(`Error while finding comment with id ${payload.id}`)
         }
+
+        const commentService = new CommentService(this.user)
```

Comment:
can this be done outside of loop above? it will improve the performance

Reply:
Changed made in 70bc2c7391eb81a8de1bad8e2208f6b1abf3713a commit